tests/unit/utils/safeList.js

```diff
@@ -0,0 +1,305 @@
+                            VersionId: 'my-deleted-version-id',
+                            IsLatest: true,
+                            LastModified: '2021-01-01T00:00:00.000Z',
+                            ETag: 'my-etag',
+                            Owner: {
+                                ID: 'my-id',
+                                DisplayName: 'my-display-name',
+                            },
+                        },
+                    ],
+                    CommonPrefix: [],
+                },
+            },
+            expectedData: {
+                Name: 'my-bucket',
+                IsTruncated: true,
+                MaxKeys: 1000,
+                NextKeyMarker: 'my-next-key-marker',
+                NextVersionIdMarker: 'my-next-version-id-marker',
+                Versions: [
+                    {
+                        Key: 'my-key',
+                        VersionId: 'my-version-id',
+                        IsLatest: true,
+                        LastModified: new Date('2021-01-01T00:00:00.000Z'),
+                        ETag: 'my-etag',
+                        Size: 'foobar',
+                        StorageClass: 'STANDARD',
+                        Owner: {
+                            ID: 'my-id',
+                            DisplayName: 'my-display-name',
+                        },
+                    },
+                ],
+                DeleteMarkers: [
+                    {
+                        Key: 'my-deleted-key',
+                        VersionId: 'my-deleted-version-id',
+                        IsLatest: true,
+                        LastModified: new Date('2021-01-01T00:00:00.000Z'),
+                        ETag: 'my-etag',
+                        Owner: {
+                            ID: 'my-id',
+                            DisplayName: 'my-display-name',
+                        },
+                    },
+                ],
+                CommonPrefixes: [],
+            },
+        },
+        {
+            description: 'should pass through original error is response is not recoverable',
+            response: {},
+            expectedError: new Error('im an original error'),
+        },
+    ];
+
+
+    testCases.forEach(testCase => {
+        it(testCase.description, done => {
+            const xmlBuilder = new xml2js.Builder();
+            const body = Buffer.from(xmlBuilder.buildObject(testCase.response));
+            const resp = {
+                httpResponse: {
+                    body,
+                },
+                error: testCase.expectedError || null,
+            };
+
+            recoverListing(resp, (err, data) => {
+                expect(err).toStrictEqual(testCase.expectedError || null);
+                expect(data).toStrictEqual(testCase.expectedData);
+                done();
+            });
+        });
+    });
+
+    it('should pass through original error if the XML fails to parse', () => {
```

Comment:
Don't you need a `done` callback to make this test async as well?

Reply:
strangely it seems to work without, but I'll add one anyway for consistency.